### PR TITLE
blog: Debian Code Search: positional index, TurboPFor-compressed

### PR DESCRIPTION
**Pull request by stapelberg**

### CODE REVIEW

File: content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+		<div class="bar" style="width: 75.6%; background-color: blue; color: white">
+			&nbsp;
+		</div>
+	</div>
+</div>
+<span style="margin-right: 2em">&nbsp;&nbsp;93G</span>
+positional (posrel)
+
+
+As you can see, the posrel technique reduces the index size to about ¾.
```
Comment (darkskiez) [inline content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown]: Maybe introduce this term first.

File: content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+current document.
+
+This is an easy way of saving some space without modifying the TurboPFor
+on-disk format.
+
+
+With the increase in size, the Linux page cache hit ratio will be lower for
+the positional index, i.e. more data will need to be fetched from disk for
+querying the index.
+
```
Comment (darkskiez) [inline content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown]: But the non-positional index can easily be stored in ram entirely on a moderate sized VM.

Had me wondering about some hybrid approach or index layout that helps.

File: content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+The values were measured by running `iostat -d 25` just before running
+[`bench.zsh`](https://codesearch.debian.net/research/2019-08-03-dcs-new-index/)
+on an otherwise idle system.
+
+#### Query speed
+
+Even though the positional index is larger and requires more data to be read at query time (see above), thanks to the C TurboPFor library, the 2 queries on a positional index are roughly as fast as the n queries on a non-positional index (≈4s instead of ≈3s).
+
+This is more than made up for by the combined i/o matching stage, which shrinks from ≈18.5s to ≈1.3s.
+
```
Comment (darkskiez) [inline content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown]: I dont see the 18.5 on these graphs,

Comment (darkskiez) [reply content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown]: maybe just call out the numbers  (7.1 + 11.4) in the text.

Also, maybe "combined i/o and matching stages"

the graph is labelled grep though...maybe just call it matching for consistency?

Comment (stapelberg) [reply content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown]: Done.

File: content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown (hunk 2 above)
Comment (stapelberg) [reply content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown]: Yes, and we used to keep it in RAM. But the point this article is trying to make is that bottomline, it is benefitial to use a positional index on disk over a non-positional index in RAM :)

Should that be made clearer somewhere/somehow?

File: content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown (hunk 1 above)
Comment (stapelberg) [reply content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown]: Done.

File: content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown (hunk 2 above)
Comment (stapelberg) [reply content/posts/2019-08-04-dcs-positional-turbopfor-index.markdown]: Discussed verbally: maybe just add a summary sentence to the article to confirm people’s understanding.

**Comment by stapelberg** [comment]
Now published: https://michael.stapelberg.ch/posts/2019-09-29-dcs-positional-turbopfor-index/

**Pull request closed without merging**